SpatialGDK/Source/Private/Interop/GlobalStateManager.cpp

```diff
@@ -93,97 +93,149 @@ void UGlobalStateManager::ApplyAcceptingPlayersUpdate(bool bAcceptingPlayersUpda
 	}
 }
 
-void UGlobalStateManager::LinkExistingSingletonActors()
+void UGlobalStateManager::LinkExistingSingletonActor(const UClass* SingletonActorClass)
 {
-	if (!NetDriver->IsServer())
+	const Worker_EntityId* SingletonEntityIdPtr = SingletonNameToEntityId.Find(SingletonActorClass->GetPathName());
+	if (SingletonEntityIdPtr == nullptr)
 	{
+		// No entry in SingletonNameToEntityId for this singleton class type
+		UE_LOG(LogGlobalStateManager, Log, TEXT("LinkExistingSingletonActor %s failed to find entry"), *SingletonActorClass->GetName());
 		return;
 	}
 
-	for (const auto& Pair : SingletonNameToEntityId)
+	const Worker_EntityId SingletonEntityId = *SingletonEntityIdPtr;
+	if (SingletonEntityId == SpatialConstants::INVALID_ENTITY_ID)
 	{
-		Worker_EntityId SingletonEntityId = Pair.Value;
-
 		// Singleton Entity hasn't been created yet
-		if (SingletonEntityId == SpatialConstants::INVALID_ENTITY_ID)
-		{
-			continue;
-		}
+		UE_LOG(LogGlobalStateManager, Log, TEXT("LinkExistingSingletonActor %s entity id is invalid"), *SingletonActorClass->GetName());
+		return;
+	}
 
-		AActor* SingletonActor = nullptr;
-		USpatialActorChannel* Channel = nullptr;
-		GetSingletonActorAndChannel(Pair.Key, SingletonActor, Channel);
+	auto* ActorChannelPair = NetDriver->SingletonActorChannels.Find(SingletonActorClass);
+	if (ActorChannelPair == nullptr)
+	{
+		// Dynamically spawn singleton actor if we have queued up data - ala USpatialReceiver::ReceiveActor - JIRA: 735
 
-		// Singleton wasn't found or channel is already set up
-		if (Channel == nullptr || Channel->Actor != nullptr)
-		{
-			check(Channel == nullptr || Channel->Actor == SingletonActor);
-			continue;
-		}
+		// No local actor has registered itself as replicatible on this worker
+		UE_LOG(LogGlobalStateManager, Log, TEXT("LinkExistingSingletonActor no actor registered"), *SingletonActorClass->GetName());
+		return;
+	}
 
-		SingletonActor->Role = ROLE_SimulatedProxy;
-		SingletonActor->RemoteRole = ROLE_Authority;
+	AActor*& SingletonActor = ActorChannelPair->Key;
```

Comment:
Does this need to be a reference? It looks noisy.

Reply:
Doesn't need to be, will change.